spyder/widgets/ipythonconsole/shell.py

```diff
@@ -219,27 +222,44 @@ def clear_console(self):
         else:
             self.execute("%clear")
 
-    def reset_namespace(self, force=False):
+    def reset_namespace(self, warning=None, silent=True):
         """Reset the namespace by removing all names defined by the user."""
         reset_str = _("Reset IPython namespace")
         warn_str = _("All user-defined variables will be removed."
                      "<br>Are you sure you want to reset the namespace?")
-        if not force:
-            reply = QMessageBox.question(self, reset_str,
-                                         warn_str,
-                                         QMessageBox.Yes | QMessageBox.No
-            )
-
-            if reply == QMessageBox.Yes:
-                if self._reading:
-                    self.dbg_exec_magic('reset', '-f')
-                else:
-                    self.execute("%reset -f")
+
+        if warning is None:
+            section = 'ipython_console'
+            option = 'show_reset_namespace_warning'
+            warning = CONF.get(section, option)
+
+        if warning:
+            box = MessageCheckBox(icon=QMessageBox.Warning, parent=self)
+            box.setWindowTitle(reset_str)
+            box.set_checkbox_text(_("Don't show again."))
+            box.setStandardButtons(QMessageBox.Yes | QMessageBox.No)
+            box.setDefaultButton(QMessageBox.No)
+
+            box.set_checked(False)
+            box.set_check_visible(True)
+            box.setText(warn_str)
+
+            answer = box.exec_()
+
+            # Update checkbox based on user interaction
+            CONF.set(section, option, not box.is_checked())
+
+            if answer != QMessageBox.Yes:
+                return
+
+        if self._reading:
+            self.dbg_exec_magic('reset', '-f')
         else:
-            if self._reading:
-                self.dbg_exec_magic('reset', '-f')
-            else:
+            if silent:
                 self.silent_execute("%reset -f")
+                self.sig_namespace_reseted.emit()
```

Comment:
Why the signal is only emitted when silent is True?

Reply:
When silent is False, something is written in the console, which trigger a refresh of the variable table automatically, so the signal is not needed. I was afraid that by emitting `sig_namespace_reseted` also when silent is False, we would trigger two calls to `refresh_table`.

Reply:
>I was afraid that by emitting sig_namespace_reseted also when silent is False, we would trigger two calls to refresh_table.

Hmm yes, that will happen; but I'll expect a signal named `sig_namespace_reseted` to be emited every time that the namespace is reseted

I don't know how to improve this, maybe a signal `namespace_changed` that'll be emitted by all things that change the namespace, and will be connected to `refresh_table`, although that is out of the scope of this PR

In the meanwhile we could add a comment explaining why the signal is only emitted  when silent=True

Reply:
Ok good idea, I will add a comment.